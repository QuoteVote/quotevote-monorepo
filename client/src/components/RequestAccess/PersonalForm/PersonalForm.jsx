import { makeStyles } from '@mui/material/styles'
import Grid from '@mui/material/Grid'
import Button from '@mui/material/Button'
import Typography from '@mui/material/Typography'
import Card from '@mui/material/Card'
import CardHeader from '@mui/material/CardHeader'
import CardContent from '@mui/material/CardContent'
import { TextField } from '@mui/material'
import PropTypes from 'prop-types'
import requestAccessStyles from '../requestAccessStyles'
import PaymentMethod from '../PaymentMethod/PaymentMethod'

const useStyles = makeStyles(requestAccessStyles)

const PersonalForm = (props) => {
  const classes = useStyles()
  const {
    requestInviteSuccessful,
    handleSubmit,
    isContinued,
    onContinue,
    errors,
    register,
    setCardDetails,
    cardDetails,
    onSubmit,
    errorMessage,
    loading,
  } = props

  return (
    <Grid container justify="center" style={{ marginRight: 24 }} spacing={2}>
      <Grid item xs={12}>
        <Typography align="center" className={classes.header}>
          {requestInviteSuccessful ? 'Thank you for' : 'Get access to your'}
          {' '}
          <span className={classes.header} style={{ color: '#52b274' }}>
            {requestInviteSuccessful ? 'joining us' : 'Personal Plan!'}
          </span>
        </Typography>
      </Grid>
      <Grid item xs={12} hidden={requestInviteSuccessful}>
        <Typography align="center" className={classes.subHeader}>
          Pay what you like, or pay nothing at all
        </Typography>
      </Grid>
      <Grid item xs={12} style={{ marginTop: requestInviteSuccessful ? '4%' : '2%' }}>
        <Grid container spacing={2}>
          <Grid container item xs={12} md={6} justify="center" alignItems="center">
            <img
              alt="Personal Plan"
              height={500}
              src="/assets/PersonalPlan.png"
              style={{
                width: '489px',
                height: '265px',
                objectFit: 'contain',
              }}
            />
          </Grid>
          {requestInviteSuccessful ? (
            <Grid container item xs={12} md={6} justify="center" alignItems="center">
              <div className={classes.opaqueBackground}>
                <Typography className={classes.message}>
                  <br />
                  When an account becomes available, an
                  <br />
                  invite will be sent to the email address you
                  <br />
                  provided.
                </Typography>
              </div>
            </Grid>
          ) : (
            <Grid item container xs={11} md={6} spacing={2}>
              <Grid item xs={12}>
                <Card>
                  <CardHeader
                    avatar={(
                      <Typography className={classes.stepNumber}>
                        1
                      </Typography>
                    )}
                    title={(
                      <Typography
                        style={{
                          font: 'Roboto',
                          fontsize: '18px',
                          lineHeight: 1.56,
                        }}
                      >
                        Your Personal Info
                      </Typography>
                    )}
                  />

                  {!isContinued && (
                    <form onSubmit={handleSubmit(onContinue)}>
                      <CardContent>
                        <Grid container spacing={2}>
                          <Grid item xs={6}>
                            <TextField
                              fullWidth
                              required
                              label="First Name"
                              name="firstName"
                              id="firstName"
                              error={errors.firstName}
                              helperText={errors.firstName && errors.firstName.message}
                              inputRef={register({
                                required: 'First Name is required',
                                minLength: {
                                  value: 1,
                                  message: 'First Name should be more than 1 character',
                                },
                                maxLength: {
                                  value: 20,
                                  message: 'First Name should be less than twenty characters',
                                },
                              })}
                            />
                          </Grid>
                          <Grid item xs={6}>
                            <TextField
                              fullWidth
                              required
                              label="Last Name"
                              name="lastName"
                              id="lastName"
                              error={errors.lastName}
                              helperText={errors.lastName && errors.lastName.message}
                              inputRef={register({
                                required: 'Last Name is required',
                                minLength: {
                                  value: 1,
                                  message: 'Last Name should be more than 1 character',
                                },
                                maxLength: {
                                  value: 20,
                                  message: 'Last Name should be less than twenty characters',
                                },
                              })}
                            />
                          </Grid>
                          <Grid item xs={12}>
                            <TextField
                              fullWidth
                              required
                              label="Email"
                              name="email"
                              id="email"
                              error={errors.email}
                              helperText={errors.email && errors.email.message}
                              inputRef={register({
                                required: 'Email is required',
                                pattern: {
                                  value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
                                  message: 'Invalid email address',
                                },
                              })}
                            />
                          </Grid>
                          <Grid item xs={12}>
                            <Button variant="contained" className={classes.greenBtn} type="submit">
                              Continue
                            </Button>
                          </Grid>
                        </Grid>
                      </CardContent>
                    </form>
                  )}
                </Card>
              </Grid>
              <Grid item xs={12}>
                <PaymentMethod
                  cardDetails={cardDetails}
                  onSubmit={onSubmit}
                  isContinued={isContinued}
                  setCardDetails={setCardDetails}
                  errorMessage={errorMessage}
                  loading={loading}
                />
              </Grid>
            </Grid>
          )}
        </Grid>
      </Grid>
    </Grid>
  );
}

PersonalForm.propTypes = {
  requestInviteSuccessful: PropTypes.bool.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  isContinued: PropTypes.bool.isRequired,
  onContinue: PropTypes.func.isRequired,
  errors: PropTypes.any.isRequired,
  register: PropTypes.any.isRequired,
  setCardDetails: PropTypes.func.isRequired,
  cardDetails: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired,
  errorMessage: PropTypes.any,
  loading: PropTypes.bool,
}
export default PersonalForm
