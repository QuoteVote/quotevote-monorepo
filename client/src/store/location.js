/**
 * Location Redux Store
 * Manages user's current location and permission state
 */

import { createSlice } from '@reduxjs/toolkit';
import { getCurrentPosition, checkPermission, GeoLocationError } from '../utils/geolocationService';

const initialState = {
  currentLocation: null, // { lat, lng }
  permissionState: 'prompt', // 'prompt' | 'granted' | 'denied' | 'unsupported'
  isLoading: false,
  error: null, // { type, message }
  lastUpdated: null,
};

const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    setLocation: (state, action) => {
      state.currentLocation = action.payload;
      state.lastUpdated = Date.now();
      state.error = null;
    },
    setPermissionState: (state, action) => {
      state.permissionState = action.payload;
    },
    setLoading: (state, action) => {
      state.isLoading = action.payload;
    },
    setError: (state, action) => {
      state.error = action.payload;
      state.isLoading = false;
    },
    clearError: (state) => {
      state.error = null;
    },
    clearLocation: (state) => {
      state.currentLocation = null;
      state.lastUpdated = null;
      state.error = null;
    },
    reset: () => initialState,
  },
});

export const {
  setLocation,
  setPermissionState,
  setLoading,
  setError,
  clearError,
  clearLocation,
  reset,
} = locationSlice.actions;

// Thunks

/**
 * Fetch user's current location
 * @param {object} options - Options passed to getCurrentPosition
 */
export const fetchLocation = (options = {}) => async (dispatch) => {
  dispatch(setLoading(true));
  dispatch(clearError());

  try {
    const location = await getCurrentPosition(options);
    dispatch(setLocation(location));
    dispatch(setPermissionState('granted'));
    return location;
  } catch (error) {
    dispatch(setError({
      type: error.type || GeoLocationError.UNKNOWN,
      message: error.message || 'Failed to get location',
    }));
    
    if (error.type === GeoLocationError.PERMISSION_DENIED) {
      dispatch(setPermissionState('denied'));
    }
    
    throw error;
  } finally {
    dispatch(setLoading(false));
  }
};

/**
 * Check and update permission state
 */
export const updatePermissionState = () => async (dispatch) => {
  try {
    const state = await checkPermission();
    dispatch(setPermissionState(state));
    return state;
  } catch (error) {
    console.error('Failed to check permission:', error);
    return 'prompt';
  }
};

// Selectors

export const selectLocation = (state) => state.location.currentLocation;
export const selectPermissionState = (state) => state.location.permissionState;
export const selectIsLoading = (state) => state.location.isLoading;
export const selectError = (state) => state.location.error;
export const selectLastUpdated = (state) => state.location.lastUpdated;

export const selectHasLocation = (state) => state.location.currentLocation !== null;
export const selectIsPermissionGranted = (state) => state.location.permissionState === 'granted';
export const selectIsPermissionDenied = (state) => state.location.permissionState === 'denied';

export default locationSlice.reducer;
