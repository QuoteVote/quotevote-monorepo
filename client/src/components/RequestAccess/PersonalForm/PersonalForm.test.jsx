import { vi } from 'vitest'
import React from 'react'
import { render, act } from '@testing-library/react'
import { Provider } from 'react-redux'
import PersonalForm from './PersonalForm'
import store from '../../../store/store'

// PersonalForm composes PaymentMethod which triggers async updates and
// requires several props. Mock it to a simple stub for this unit test so
// we avoid PropTypes warnings and act() timing issues.
vi.mock('../PaymentMethod/PaymentMethod', () => {
  return {
    __esModule: true,
    default: (props) => <div data-testid="mock-payment-method" {...props} />,
  }
})

const props = {
  handleSubmit: vi.fn(),
  register: vi.fn(),
  errors: {
    fullName: '',
  },
  requestInviteSuccessful: false,
}

const PersonalFormWrapper = () => (
  <Provider store={store}>
    <PersonalForm {...props} />
  </Provider>
)

describe('PersonalForm test -', () => {
  it('renders correctly', async () => {
    let container
    await act(async () => {
      const res = render(<PersonalFormWrapper />)
      container = res.container
    })
    expect(container.firstChild).toMatchSnapshot()
  })
})
