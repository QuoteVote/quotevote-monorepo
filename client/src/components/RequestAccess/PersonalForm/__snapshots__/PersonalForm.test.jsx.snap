// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PersonalForm test - > renders correctly 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root"
  justify="center"
  style="margin-right: 24px;"
>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter makeStyles-header-2 css-d4fp74-MuiTypography-root"
    >
      Get access to your
       
      <span
        class="makeStyles-header-2"
        style="color: rgb(82, 178, 116);"
      >
        Personal Plan!
      </span>
    </p>
  </div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter makeStyles-subHeader-3 css-d4fp74-MuiTypography-root"
    >
      Pay what you like, or pay nothing at all
    </p>
  </div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
    style="margin-top: 2%;"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row css-1rzj21q-MuiGrid-root"
        justify="center"
      >
        <img
          alt="Personal Plan"
          height="500"
          src="/assets/PersonalPlan.png"
          style="width: 489px; height: 265px; object-fit: contain;"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-yu7yff-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 makeStyles-stepNumber-4 css-1edfpdg-MuiTypography-root"
                >
                  1
                </p>
              </div>
              <div
                class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
                  style="font: Roboto; line-height: 1.56;"
                >
                  Your Personal Info
                </p>
              </div>
            </div>
            <form>
              <div
                class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1htra48-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="firstName"
                        id="firstName-label"
                      >
                        First Name
                        <span
                          aria-hidden="true"
                          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                        >
                           
                          *
                        </span>
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1bz4o94-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiOutlinedInput-input css-165z7mv-MuiInputBase-input-MuiOutlinedInput-input"
                          id="firstName"
                          name="firstName"
                          required=""
                          type="text"
                          value=""
                        />
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1cs9q92-MuiNotchedOutlined-root"
                          >
                            <span>
                              First Name
                               
                              *
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1htra48-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="lastName"
                        id="lastName-label"
                      >
                        Last Name
                        <span
                          aria-hidden="true"
                          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                        >
                           
                          *
                        </span>
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1bz4o94-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiOutlinedInput-input css-165z7mv-MuiInputBase-input-MuiOutlinedInput-input"
                          id="lastName"
                          name="lastName"
                          required=""
                          type="text"
                          value=""
                        />
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1cs9q92-MuiNotchedOutlined-root"
                          >
                            <span>
                              Last Name
                               
                              *
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1htra48-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="email"
                        id="email-label"
                      >
                        Email
                        <span
                          aria-hidden="true"
                          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                        >
                           
                          *
                        </span>
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1bz4o94-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiOutlinedInput-input css-165z7mv-MuiInputBase-input-MuiOutlinedInput-input"
                          id="email"
                          name="email"
                          required=""
                          type="text"
                          value=""
                        />
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1cs9q92-MuiNotchedOutlined-root"
                          >
                            <span>
                              Email
                               
                              *
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary makeStyles-greenBtn-1 css-g1oxnb-MuiButtonBase-root-MuiButton-root"
                      tabindex="0"
                      type="submit"
                    >
                      Continue
                    </button>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
        >
          <form>
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
            >
              <div
                class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
              >
                <div
                  class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 makeStyles-stepNumber-24 css-1edfpdg-MuiTypography-root"
                  >
                    2
                  </p>
                </div>
                <div
                  class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 makeStyles-stepName-25 css-1edfpdg-MuiTypography-root"
                  >
                    Would you like to support this app?
                  </p>
                </div>
              </div>
              <input
                name="cardNumber"
                type="hidden"
              />
              <input
                name="expiry"
                type="hidden"
              />
              <input
                name="cvv"
                type="hidden"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
